modules/dolt/dolt.go

```diff
@@ -0,0 +1,250 @@
+		return
+	}
+
+	// create database
+	_, err = db.Exec(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %s;", c.database))
+	if err != nil {
+		err = fmt.Errorf("error creating database %s: %w\n", c.database, err)
+		return
+	}
+
+	if createUser {
+		// create user
+		_, err = db.Exec(fmt.Sprintf("CREATE USER IF NOT EXISTS '%s' IDENTIFIED BY '%s';", c.username, c.password))
+		if err != nil {
+			err = fmt.Errorf("error creating user %s: %w\n", c.username, err)
+			return
+		}
+
+		q := fmt.Sprintf("GRANT ALL ON %s.* TO '%s';", c.database, c.username)
+		// grant user privileges
+		_, err = db.Exec(q)
+		if err != nil {
+			err = fmt.Errorf("error creating user %s: %w\n", c.username, err)
+			return
+		}
+	}
+
+	return
+}
+
+func (c *DoltContainer) initialConnectionString(ctx context.Context) (string, error) {
+	containerPort, err := c.MappedPort(ctx, "3306/tcp")
+	if err != nil {
+		return "", err
+	}
+
+	host, err := c.Host(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	connectionString := fmt.Sprintf("root:@tcp(%s:%s)/", host, containerPort.Port())
+	return connectionString, nil
+}
+
+func (c *DoltContainer) ConnectionString(ctx context.Context, args ...string) (string, error) {
+	containerPort, err := c.MappedPort(ctx, "3306/tcp")
+	if err != nil {
+		return "", err
+	}
+
+	host, err := c.Host(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	extraArgs := ""
+	if len(args) > 0 {
+		extraArgs = strings.Join(args, "&")
+	}
+	if extraArgs != "" {
+		extraArgs = "?" + extraArgs
+	}
+
+	connectionString := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s%s", c.username, c.password, host, containerPort.Port(), c.database, extraArgs)
+	return connectionString, nil
+}
+
+func WithUsername(username string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["DOLT_USER"] = username
+	}
+}
+
+func WithPassword(password string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["DOLT_PASSWORD"] = password
+	}
+}
+
+func WithDoltCredsPublicKey(key string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["DOLT_CREDS_PUB_KEY"] = key
+	}
+}
+
+func WithDoltCloneRemoteUrl(url string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["DOLT_REMOTE_CLONE_URL"] = url
+	}
+}
+
+func WithDatabase(database string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["DOLT_DATABASE"] = database
+	}
+}
+
+func WithConfigFile(configFile string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		cf := testcontainers.ContainerFile{
+			HostFilePath:      configFile,
+			ContainerFilePath: "/etc/dolt/servercfg.d/server.cnf",
+			FileMode:          0o755,
+		}
+		req.Files = append(req.Files, cf)
+	}
+}
+
```

Comment:
Do you think this is a good candidate for a `WithNewBranch(branch string)` functional option? https://github.com/dolthub/testcontainers-go-demo/blob/6556db5866f785c4f35f51370036093f7710d3e6/customer/repo_suite_test.go#L59C2-L63C3

Reply:
I'm not sure this is a good candidate to add at the moment, since I think it would be better handled either in the the init script with `CALL DOLT_CHECKOUT()` used to create a new branch, and then that branch can be checked out in test harness itself, like I've done in the demo repo. Doing this in the script or in the harness allows for a lot of flexibility with branch navigation and management, which I think is a good place to start (since database branches are a novel concept to most everyone anyway).

Adding `WithNewBranch(branch string)` could be used to create a new branch from the default branch during Dolt container initialization, but then the connection string returned from `ConnectionString()` would also need to have the new branch name appended to it in the form `user:pass@tcp(host:port)/dbname/newBranchName` in order for new sessions to connect to that branch by default. Otherwise, new sessions would _still_ connect to the default branch. I wouldn't want people to think that using `WithNewBranch(branch string)` means that all new database sessions occur against this new branch. But it's possible this is behavior people want, I'm not sure.

If this isn't a merge blocker, I'm happy to add that later upon request.

Reply:
It would be advantageous to have the current connection string prepared in a way that includes all the branch and commit information. This would eliminate the need for users to manually concatenate strings, allowing them to connect effortlessly to the pre-prepared state. 

Maybe with a `container.CurrentConnectionString(ctx)`. But I am not know what would be the best solution.

Reply:
or `container.ConnectionStringWithBranch(ctx)` or `ConnectionStringWithCommit` / `ConnectionStringWithTag`

Reply:
I like this conversation as it's building a "roadmap" for the module :)

So yes, not a blocker on my side, you're the experts 😊 , so we can have follow-up PRs for it.